File: outbreak_cleaned.py
import pandas as pd
import numpy as np
from scipy.stats import chi2_contingency, fisher_exact
import statsmodels.api as sm
import statsmodels.formula.api as smf
import matplotlib.pyplot as plt

# 1. DATA LOADING AND INITIAL PREPARATION ----

# Load the data
df = pd.read_csv("fake_outbreak_dataset.csv")

# Convert all columns except 'personID' and the dates to 'category' dtype
# Define a list of columns to exclude from being converted to 'category' data type
exclude_columns = ['personID', 'onset_date', 'exposure_date']

# Loop through all columns and convert them to 'category' data type except the excluded ones
for col in df.columns:
    if col not in exclude_columns:
        df[col] = df[col].astype('category')

        
# Display a summary (Python equivalent does not directly compute frequencies like R's summary)
print(df.describe(include='all'))

# 2. ATTACK RATE ANALYSIS ----

def calculate_attack_rate(data, item):
    grouped = data.groupby(item).agg(Total=('illness', 'count'),
                                      Ill=('illness', lambda x: (x == 'Y').sum()))
    grouped['Attack_Rate'] = grouped['Ill'] / grouped['Total'] * 100
    grouped.reset_index(inplace=True)
    grouped['Item'] = item  # Add the item as a column for identification
    return grouped[['Item', item, 'Total', 'Ill', 'Attack_Rate']]

items = df.columns[2:11]  # Assuming 'personID' and 'illness' are first two columns

# Using a list comprehension and Pandas concat to mimic R's map_df functionality
attack_rates = pd.concat([calculate_attack_rate(df, item) for item in items], ignore_index=True)
print(attack_rates)

# 3. STATISTICAL ANALYSIS ----

def perform_chi_square(data, dependent_var):
    results = []
    factor_vars = data.select_dtypes(include=['category']).columns.drop([dependent_var])

    for var in factor_vars:
        print(f"\nAnalyzing: {var} against {dependent_var}")
        
        cross_tab = pd.crosstab(data[var], data[dependent_var])
        print("Cross Table:\n", cross_tab)
        
        chi2, p, dof, expected = chi2_contingency(cross_tab)
        print(f"\nChi-square Test:\nChi2: {chi2}, p-value: {p}")
        
        # Fisher's Exact Test and odds ratio calculation for 2x2 tables
        if cross_tab.shape == (2, 2):
            odds_ratio, p_value = fisher_exact(cross_tab)
            print(f"Odds Ratio: {odds_ratio}, p-value (Fisher's Exact): {p_value}")
        else:
            odds_ratio = np.nan
            print("Odds Ratio not calculated, table not 2x2.")
        
        results.append({
            'Variable': var,
            'Odds_Ratio': odds_ratio,
            'X_Squared_Statistic': chi2,
            'p_Value': p
        })
        
    return pd.DataFrame(results)

results_summary = perform_chi_square(df, 'illness')
print(results_summary)

# 4. EPI CURVE ----

# Convert 'onset_date' to datetime format
df['onset_date'] = pd.to_datetime(df['onset_date'])

# Make the epi curve
plt.figure(figsize=(10, 6))
# Calculate the number of bins based on the range of dates and desired width of each bin (21600 seconds or 6 hours here)
bin_width_seconds = 21600  # 6 hours in seconds
num_bins = int((df['onset_date'].max() - df['onset_date'].min()).total_seconds() / bin_width_seconds)

sns.histplot(data=df, x='onset_date', bins=num_bins)
plt.title("Epidemiological Curve of Onset Dates")
plt.xlabel("Onset Date")
plt.ylabel("Number of Cases")
plt.xticks(rotation=90)
plt.show()
plt.clf()

File: outbreak_cleaned.R
# Load necessary libraries
library(tidyverse)
library(gt)

# 1. DATA LOADING AND INITIAL PREPARATION ----

# Load the data from a CSV file into a dataframe called 'df'
df <- read_csv("fake_outbreak_dataset.csv")

# Convert all columns except 'personID' to factor data type
# The '[,-1]' selects all columns except the first one (assuming 'personID' is the first)
df <- df %>% 
  mutate(across(-personID, as.factor)) 

# Display a summary of the dataframe, showing frequencies of all variables
summary(df)

# 2. ATTACK RATE ANALYSIS ----

# Define a function to calculate attack rates for given items
calculate_attack_rate <- function(data, item) {
  data %>%
    group_by(!!sym(item)) %>%
    summarise(Total = n(),
              Ill = sum(illness == 'Y'),
              Attack_Rate = Ill / Total * 100) %>% # Calculate as percentage
    mutate(Item = item) %>% # Add the item as a column
    select(Item, !!sym(item), Total, Ill, Attack_Rate) # Select and order columns
}

# Define the items (variables) to calculate attack rates for, excluding 'personID' and another column
items <- names(df)[3:ncol(df)]

# Calculate attack rates for each item and combine the results into one dataframe
attack_rates <- map_df(items, calculate_attack_rate, data = df)

# Prettier table
attack_rates %>%
  gt() %>%
  tab_header(
    title = "Attack Rates",
    subtitle = "Calculated attack rates for various items"
  ) %>%
  fmt_integer(
    columns = vars(Attack_Rate)
  ) %>%
  cols_label(
    Item = "Item",
    Total = "Total Cases",
    Ill = "Number Ill",
    Attack_Rate = "Attack Rate (%)"
  )


# Display the attack rate results
print(attack_rates)

# 3. STATISTICAL ANALYSIS ----

# Define a function to perform chi-square tests and calculate odds ratio for a dependent variable against all factor variables
perform_chi_square <- function(data, dependent_var) {
  
  # Identify all factor variables excluding the dependent variable
  factor_vars <- names(select(data, where(is.factor))) %>% setdiff(dependent_var) 
  
  results_summary <- purrr::map(factor_vars, function(var) {
    cat("\nAnalyzing:", var, "against", dependent_var, "\n")
    
    cross_tab <- table(data[[var]], data[[dependent_var]])
    
    cat("Cross Table:\n")
    print(cross_tab)
    
    # Perform Fisher's Exact Test for 2x2 tables to calculate Odds Ratio
    if (all(dim(cross_tab) == c(2, 2))) {
      odds_ratio_test <- fisher.test(cross_tab)
      odds_ratio <- odds_ratio_test$estimate
      conf_interval <- odds_ratio_test$conf.int
      cat("Odds Ratio:", odds_ratio, "\n")
      cat("95% Confidence Interval for Odds Ratio:", conf_interval[1], "to", conf_interval[2], "\n")
    } else {
      odds_ratio <- NA
      conf_interval <- c(NA, NA)
      cat("Odds Ratio and Confidence Interval not calculated, table not 2x2.\n")
    }
    
    # Perform Chi-square test
    chi_test_result <- chisq.test(cross_tab)
    cat("\nChi-square Test:\n")
    print(chi_test_result)
    
    # Compile test results into a list
    list(
      Variable = var,
      Odds_Ratio = odds_ratio,
      Conf_Interval_From = conf_interval[1],
      Conf_Interval_To = conf_interval[2],
      X_Squared_Statistic = chi_test_result$statistic,
      p_Value = chi_test_result$p.value
    )
  }) %>% bind_rows() # Combine all lists into a dataframe
  
  # Return the compiled results
  results_summary
}

# Execute the chi-square analysis function for each variable against "illness" and print results
results_summary <- perform_chi_square(df, "illness")
print(results_summary)
